# CPSC 330 - Applied Machine Learning 

## Homework 3: Preprocessing 
### Associated lectures: [Lectures 4, 5, 6](https://ubc-cs.github.io/cpsc330/README.html) 

**Due date: Monday, Oct 04, 2021 at 11:59pm**

## Table of Contents

- [Instructions](#si)
- [Introduction](#in)
- [Exercise 1: Introducing the dataset](#1)
- [Exercise 2: Exploratory data analysis (EDA)](#2)
- [Exercise 3: Preprocessing](#3)
- [Exercise 4: Building models](#4)
- [Exercise 5: Evaluating on the test set](#5)

## Imports 

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.compose import ColumnTransformer, make_column_transformer
from sklearn.dummy import DummyClassifier
from sklearn.impute import SimpleImputer
from sklearn.model_selection import cross_val_score, cross_validate, train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from pandas_profiling import ProfileReport

## Instructions <a name="si"></a>
<hr>
rubric={points:6}

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330/blob/master/docs/homework_instructions.md). 

**You may work with a partner on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).

_Note: The assignments will get gradually more open-ended as we progress through the course. In many cases, there won't be a single correct solution. Sometimes you will have to make your own choices and your own decisions (for example, on what parameter values to use when they are not explicitly provided in the instructions). Use your own judgment in such cases and justify your choices, if necessary._

## Introduction <a name="in"></a>
<hr>

A crucial step when using machine learning algorithms on real-world datasets is preprocessing. This homework will give you some practice of data preprocessing and building a supervised machine learning pipeline on a real-world dataset. 

<br><br>

## Exercise 1: Introducing the dataset <a name="1"></a>
<hr>

In this lab, you will be working on [the adult census dataset](https://www.kaggle.com/uciml/adult-census-income#). Download the CSV and save it as `adult.csv` locally in this homework folder. 

This is a classification dataset and the classification task is to predict whether income exceeds 50K per year or not based on the census data. You can find more information on the dataset and features [here](http://archive.ics.uci.edu/ml/datasets/Adult).

The starter code below loads the data CSV (assuming that it is saved as `adult.csv` in this folder). 

_Note that many popular datasets have sex as a feature where the possible values are male and female. This representation reflects how the data were collected and is not meant to imply that, for example, gender is binary._

In [2]:
census_df = pd.read_csv("adult.csv")
census_df.shape

(32561, 15)

<br><br>

### 1.1 Data splitting 
rubric={points:4}

In order to avoid violation of the golden rule, the first step before we do anything is splitting the data. 

**Your tasks:**

1. Split the data into `train_df` (60%) and `test_df` (40%) with `random_state = 42`. Keep the target column (`income`) in the splits so that we can use it in the exploratory data analysis.  

_Usually having more data for training is a good idea. But here I'm using 60%/40% split because this is kind of a big dataset for a modest laptop. A smaller training data means it won't take too long to train the model on your laptop. A side advantage of this would be that with a bigger test split, we'll have a more reliable estimate of the deployment performance!_

In [3]:
train_df, test_df = train_test_split(census_df, test_size=0.4,random_state=42)

<br><br>

## Exercise 2: Exploratory data analysis (EDA) <a name="2"></a> 
<hr>

Let's examine our `train_df`. 

In [4]:
train_df.sort_index()

age workclass  fnlwgt     education  education.num      marital.status  \
0       90         ?   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66         ?  186061  Some-college             10             Widowed   
5       34   Private  216864       HS-grad              9            Divorced   
9       41   Private   70037  Some-college             10       Never-married   
...    ...       ...     ...           ...            ...                 ...   
32553   43   Private   84661     Assoc-voc             11  Married-civ-spouse   
32554   32   Private  116138       Masters             14       Never-married   
32555   53   Private  321865       Masters             14  Married-civ-spouse   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   

            occupation   relationship                race     sex  \
0                    ?  Not-in-family               White  Female   
1      Exec-managerial  Not-in-family               White  Female   
2                    ?      Unmarried               Black  Female   
5        Other-service      Unmarried               White  Female   
9         Craft-repair      Unmarried               White    Male   
...                ...            ...                 ...     ...   
32553            Sales        Husband               White    Male   
32554     Tech-support  Not-in-family  Asian-Pac-Islander    Male   
32555  Exec-managerial        Husband               White    Male   
32557     Tech-support           Wife               White  Female   
32559     Adm-clerical      Unmarried               White  Female   

       capital.gain  capital.loss  hours.per.week native.country income  
0                 0          4356              40  United-States  <=50K  
1                 0          4356              18  United-States  <=50K  
2                 0          4356              40  United-States  <=50K  
5                 0          3770              45  United-States  <=50K  
9                 0          3004              60              ?   >50K  
...             ...           ...             ...            ...    ...  
32553             0             0              45  United-States  <=50K  
32554             0             0              11         Taiwan  <=50K  
32555             0             0              40  United-States   >50K  
32557             0             0              38  United-States  <=50K  
32559             0             0              40  United-States  <=50K  

[19536 rows x 15 columns]

We see some missing values represented with a "?". Probably these were the questions not answered by some people during the census.  Usually `.describe()` or `.info()` methods would give you information on missing values. But here, they won't pick "?" as missing values as they are encoded as strings instead of an actual NaN in Python. So let's replace them with `np.nan` before we carry out EDA. If you do not do it, you'll encounter an error later on when you try to pass this data to a classifier. 

In [5]:
train_df_nan = train_df.replace("?", np.nan)
test_df_nan = test_df.replace("?", np.nan)
train_df_nan.shape

(19536, 15)

In [6]:
train_df_nan.sort_index()

age workclass  fnlwgt     education  education.num      marital.status  \
0       90       NaN   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66       NaN  186061  Some-college             10             Widowed   
5       34   Private  216864       HS-grad              9            Divorced   
9       41   Private   70037  Some-college             10       Never-married   
...    ...       ...     ...           ...            ...                 ...   
32553   43   Private   84661     Assoc-voc             11  Married-civ-spouse   
32554   32   Private  116138       Masters             14       Never-married   
32555   53   Private  321865       Masters             14  Married-civ-spouse   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   

            occupation   relationship                race     sex  \
0                  NaN  Not-in-family               White  Female   
1      Exec-managerial  Not-in-family               White  Female   
2                  NaN      Unmarried               Black  Female   
5        Other-service      Unmarried               White  Female   
9         Craft-repair      Unmarried               White    Male   
...                ...            ...                 ...     ...   
32553            Sales        Husband               White    Male   
32554     Tech-support  Not-in-family  Asian-Pac-Islander    Male   
32555  Exec-managerial        Husband               White    Male   
32557     Tech-support           Wife               White  Female   
32559     Adm-clerical      Unmarried               White  Female   

       capital.gain  capital.loss  hours.per.week native.country income  
0                 0          4356              40  United-States  <=50K  
1                 0          4356              18  United-States  <=50K  
2                 0          4356              40  United-States  <=50K  
5                 0          3770              45  United-States  <=50K  
9                 0          3004              60            NaN   >50K  
...             ...           ...             ...            ...    ...  
32553             0             0              45  United-States  <=50K  
32554             0             0              11         Taiwan  <=50K  
32555             0             0              40  United-States   >50K  
32557             0             0              38  United-States  <=50K  
32559             0             0              40  United-States  <=50K  

[19536 rows x 15 columns]

The "?" symbols are now replaced with NaN values. 

<br><br>

### 2.1 Visualizing features
rubric={points:10}

**Your tasks:**

1. Examine the information given by `train_df_nan.info()` and `train_df_nan.describe()` methods. In case of `.describe()`, use the `include="all"` argument to show summary statistics of all  features.
2. Visualize the histograms of numeric features. 
3. From the visualizations, which features seem relevant for the given prediction task?

> Note: (Optional) If you're feeling excited about this you are welcome to use [`pandas_profiling`](https://github.com/pandas-profiling/pandas-profiling) for more elaborate visualization and EDA. 

In [7]:
train_df_nan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19536 entries, 25823 to 23654
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             19536 non-null  int64 
 1   workclass       18428 non-null  object
 2   fnlwgt          19536 non-null  int64 
 3   education       19536 non-null  object
 4   education.num   19536 non-null  int64 
 5   marital.status  19536 non-null  object
 6   occupation      18424 non-null  object
 7   relationship    19536 non-null  object
 8   race            19536 non-null  object
 9   sex             19536 non-null  object
 10  capital.gain    19536 non-null  int64 
 11  capital.loss    19536 non-null  int64 
 12  hours.per.week  19536 non-null  int64 
 13  native.country  19187 non-null  object
 14  income          19536 non-null  object
dtypes: int64(6), object(9)
memory usage: 2.4+ MB


In [8]:
train_df_nan.describe(include="all")

age workclass        fnlwgt education  education.num  \
count   19536.000000     18428  1.953600e+04     19536   19536.000000   
unique           NaN         8           NaN        16            NaN   
top              NaN   Private           NaN   HS-grad            NaN   
freq             NaN     13570           NaN      6256            NaN   
mean       38.592547       NaN  1.892662e+05       NaN      10.084767   
std        13.638971       NaN  1.049020e+05       NaN       2.580723   
min        17.000000       NaN  1.228500e+04       NaN       1.000000   
25%        28.000000       NaN  1.177670e+05       NaN       9.000000   
50%        37.000000       NaN  1.782835e+05       NaN      10.000000   
75%        47.000000       NaN  2.368860e+05       NaN      12.000000   
max        90.000000       NaN  1.455435e+06       NaN      16.000000   

            marital.status      occupation relationship   race    sex  \
count                19536           18424        19536  19536  19536   
unique                   7              14            6      5      2   
top     Married-civ-spouse  Prof-specialty      Husband  White   Male   
freq                  9025            2490         7937  16676  13079   
mean                   NaN             NaN          NaN    NaN    NaN   
std                    NaN             NaN          NaN    NaN    NaN   
min                    NaN             NaN          NaN    NaN    NaN   
25%                    NaN             NaN          NaN    NaN    NaN   
50%                    NaN             NaN          NaN    NaN    NaN   
75%                    NaN             NaN          NaN    NaN    NaN   
max                    NaN             NaN          NaN    NaN    NaN   

        capital.gain  capital.loss  hours.per.week native.country income  
count   19536.000000  19536.000000    19536.000000          19187  19536  
unique           NaN           NaN             NaN             41      2  
top              NaN           NaN             NaN  United-States  <=50K  
freq             NaN           NaN             NaN          17517  14841  
mean     1090.443540     86.537162       40.532606            NaN    NaN  
std      7449.700833    402.395668       12.406636            NaN    NaN  
min         0.000000      0.000000        1.000000            NaN    NaN  
25%         0.000000      0.000000       40.000000            NaN    NaN  
50%         0.000000      0.000000       40.000000            NaN    NaN  
75%         0.000000      0.000000       45.000000            NaN    NaN  
max     99999.000000   4356.000000       99.000000            NaN    NaN

<br><br>

### 2.2 Identify transformations to apply
rubric={points:18}

**Your tasks:**
1. Identify the sequence of transformations that you would apply on each column in the dataset and fill in the table below accordingly. An example of the sequence of transformations to be applied on the `occupation` feature is shown in the table below. You may decide not to apply any transformations on a certain column or entirely drop a column from your model. That's totally fine. 
2. Are there common transformations you would like to apply on certain types of features? Identify different feature types for applying different transformations. In particular, fill in the lists below. 
3. Is including the `race` feature for predicting income ethically a good idea? Briefly discuss. 

> Note: This question is a bit open-ended and there is no single correct solution.

| Feature | Transformation |
| --- | ----------- |
| occupation | imputation, one-hot encoding |
| age | standard scaling|
| workclass |imputation, one-hot encodiing |
| fnlwgt | standard scaling|
| education | drop |
| education.num | none|
| marital.status | one-hot encoding |
| relationship | one-hot encoding |
| race | drop |
| sex | drop |
| capital.gain | standard scaling |
| capital.loss | standard scaling |
| hours.per.week | standard scaling |
| native.country | imputation, one-hot encoding|


In [9]:
# Fill in the lists below. 
# It's OK to keep some of the lists empty or add new lists. 
numeric_features = ["age", "fnlwgt", "capital.gain", "capital.loss", "hours.per.week"]
categorical_features = ["occupation", "workclass", "native.country", "marital.status", "relationship"]
ordinal_features = []
binary_features = []
drop_features = ["race", "sex", "education"]
passthrough_features = ["education.num"]
target = "income"

### For this project, I don't believe it's ethical to include race (or sex) as a predictor for income. This is because using it as a predictor can cement past disparities into models that may be used in production in the future, which can further exacerbate those economic disparities. Real examples of this include the algorithms that determine home mortgage eligibility and applicant tracking systems. My favourite commentary on this is by data scientist and author Cathy O'Neil in her book "Weapons of Math Destruction". 

### "Big Data processes codify the past. They do not invent the future. Doing that requires moral imagination, and that’s something only humans can provide". 

<br><br>

### 2.3 Separating feature vectors and targets  
rubric={points:4}

**Your tasks:**

1. Create `X_train`, `y_train`, `X_test`, `y_test` from `train_df_nan` and `test_df_nan`. 
2. At this point, if you train [`sklearn`'s `SVC`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) model on `X_train` and `y_train` would it work? Why or why not?

In [10]:
X_train = train_df_nan.drop(columns=["income"])
y_train = train_df_nan["income"]

X_test = test_df_nan.drop(columns=["income"])
y_test = test_df_nan["income"]

In [11]:
#This does not work. 
#model = SVC(gamma=0.001, C=100.)
#model.fit(X_train, y_train) 

### We can not train sklearn's SVC model on the training data yet because we are using strings to differentiate our categorical variables instead of using a numerical encoding (one-hot, ordinal). Thus, we have to preprocess the data first into a dataframe that exclusively uses numbers as "Xs". 

<br><br><br><br>

## Exercise 3: Preprocessing <a name="3"></a>
<hr>

### 3.1 Preprocessing using `sklearn`'s `ColumnTransformer` and `Pipeline`
rubric={points:18}

Let's carry out preprocessing using `sklearn`'s `ColumnTransformer` and `Pipeline`. Note that you can define pipelines in two ways: 
- by using [`Pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) and explicitly providing named steps
- by using [`make_pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.make_pipeline.html#sklearn.pipeline.make_pipeline), which automatically names the steps in the pipeline with their class names. 

Similarly you can create a column transformer in two ways:
- by using [`ColumnTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html)
- by using [`make_column_transformer`](https://scikit-learn.org/stable/modules/generated/sklearn.compose.make_column_transformer.html) 

You may use the method of your choice but `make_pipeline` and `make_column_transformer` are highly recommended.  

**Your tasks:**

1. Create a column transformer `preprocessor` based on transformations you want to apply on the data from 2.2. 
2. Transform the data by calling `fit_transform` on the training set. What's the shape of the transformed data? 
3. Why do we need to use a column transformer in this case? Briefly explain. 

In [12]:
preprocessor = make_column_transformer(
    (StandardScaler(), numeric_features), # scaling on numeric features
    (
        make_pipeline(SimpleImputer(strategy="most_frequent"), OneHotEncoder(handle_unknown="ignore")),
        categorical_features,
    ),  # Imputation with the mode and OHE on categorical features
    ("passthrough", passthrough_features),  # no transformations on the education.num
    ("drop", drop_features),  # drop the drop features
)

preprocessor

ColumnTransformer(transformers=[('standardscaler', StandardScaler(),
                                 ['age', 'fnlwgt', 'capital.gain',
                                  'capital.loss', 'hours.per.week']),
                                ('pipeline',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore'))]),
                                 ['occupation', 'workclass', 'native.country',
                                  'marital.status', 'relationship']),
                                ('passthrough', 'passthrough',
                                 ['education.num']),
                                ('drop', 'drop', ['race', 'sex', 'education'])])

In [13]:
preprocessor.fit_transform(X_train)

<19536x82 sparse matrix of type '<class 'numpy.float64'>'
	with 214896 stored elements in Compressed Sparse Row format>

### The shape of the sparse data is a sparse matrix with 19536 observations and 82 rows. 

### We need to use a column transformer because the different column classes need to be transformed differently. For example, we only apply  scaling on numeric features, imputation and one-hot encoding on categorical features, and other features need to be dropped or passed through entirely. With pipelines, all columns are transformed the same way, which is not what we're looking for in this case. 

<br><br><br><br>

## Exercise 4: Building models <a name="4"></a>
<hr>

Now that we have preprocessed features, we are ready to build models. Below, I'm providing the function we used in class which returns mean cross-validation score along with standard deviation for a given model. Feel free to use it to keep track of your results if you like. 

In [14]:
results_dict = {} # dictionary to store all the results

In [15]:

def mean_std_cross_val_scores(model, X_train, y_train, **kwargs):
    """
    Returns mean and std of cross validation

    Parameters
    ----------
    model :
        scikit-learn model
    X_train : numpy array or pandas DataFrame
        X in the training data
    y_train :
        y in the training data

    Returns
    ----------
        pandas Series with mean scores from cross_validation
    """

    scores = cross_validate(model, X_train, y_train, **kwargs)

    mean_scores = pd.DataFrame(scores).mean()
    std_scores = pd.DataFrame(scores).std()
    out_col = []

    for i in range(len(mean_scores)):
        out_col.append((f"%0.3f (+/- %0.3f)" % (mean_scores[i], std_scores[i])))

    return pd.Series(data=out_col, index=mean_scores.index)

<br><br>

### 4.1 Baseline model 
rubric={points:6}

**Your tasks:**
1. Define a pipeline with two steps: `preprocessor` from 3.1 and `scikit-learn`'s `DummyClassifier` with `strategy="prior"` as your classifier.  
2. Carry out 5-fold cross-validation with the pipeline. Store the results in `results_dict` above. Display the results as a pandas DataFrame.  

> You may use the function `mean_std_cross_val_scores` above to carry out cross-validation and storing results. Refer to the class notes if you are unsure about how to use it. 

In [16]:
pipe = make_pipeline(preprocessor, DummyClassifier(strategy="prior"))


In [17]:
 results_dict["DummyClassifier"] = mean_std_cross_val_scores(
     pipe, X_train, y_train, return_train_score=True
 )
    
    

In [18]:
pd.DataFrame(results_dict).T

fit_time         score_time         test_score  \
DummyClassifier  0.043 (+/- 0.003)  0.015 (+/- 0.001)  0.760 (+/- 0.000)   

                       train_score  
DummyClassifier  0.760 (+/- 0.000)

<br><br>

### 4.2 Trying different classifiers
rubric={points:14}

**Your tasks:**

1. For each of the models in the starter code below: 
    - Define a pipeline with two steps: `preprocessor` from 3.1 and the model as your classifier. 
    - Carry out 5-fold cross-validation with the pipeline.  
    - Store the results in `results_dict`. 
2. Display all the results so far as a pandas dataframe. 
3. Compare the train and validation accuracies and `fit` and `score` times in each case. How do the the validation accuracies compare to the baseline model from 4.1? Which model has the best validation accuracy? Which model is the fastest one?  

> Note that this might take a while to run.

> You may use the function above `mean_std_cross_val_scores` to carry out cross-validation and storing results. Refer to the class notes if you are unsure about how to use it. 

In [19]:
models = {
    "decision tree": DecisionTreeClassifier(),
    "kNN": KNeighborsClassifier(),
    "RBF SVM": SVC(),
}

In [ ]:
for key, value in models.items(): 
    print(key)
    pipe = make_pipeline(preprocessor, value)
    results_dict[key] = mean_std_cross_val_scores(
    pipe, X_train, y_train, return_train_score=True
 )
        
     

decision tree
kNN
RBF SVM


In [ ]:
pd.DataFrame(results_dict).T

### All three models outperformed the basic DummyClassifier model. The validation accuracies are all better than the DummyClassifier, with the RBF SVM having the best validation accuracy. However, the fastest model was the decision tree classifier, with an overall time of 0.498 in comparison to 2.432 and 5.929 respectively. 

<br><br>

### (optional) 4.3 Exploring importance of scaling
rubric={points:1}

In this exercise you'll examine whether scaling helps in case of KNNs and SVM RBFs. 

**Your tasks:**

1. Create a column transformer without the `StandardScaler` step for `numeric_features`. 
2. Repeat the steps in 4.2 with this new column transformer. 
3. Compare the results of scaled numeric features with unscaled numeric features. Is scaling necessary for decision trees? Why or why not?

In [ ]:
passthrough_features2 = numeric_features + passthrough_features

results_dict_without_scaling = {}

preprocessor_without_scaling = make_column_transformer(
    (
        make_pipeline(SimpleImputer(strategy="most_frequent"), OneHotEncoder(handle_unknown="ignore")),
        categorical_features,
    ),  # Imputation with the mode and OHE on categorical features
    ("passthrough", passthrough_features2),  # no transformations on the education.num
    ("drop", drop_features),  # drop the drop features
)

for key, value in models.items(): 
    print(key)
    pipe = make_pipeline(preprocessor_without_scaling, value)
    results_dict_without_scaling[key] = mean_std_cross_val_scores(
    pipe, X_train, y_train, return_train_score=True
 )

pd.DataFrame(results_dict_without_scaling).T



### The kNN and RBF test scores are considerably worse than the previously when scaling is applied. These models are effected by the scale - kNNs and radial basis functions take (euclidian) distance into consideration. However, scaling is not necessary for decision trees because there is no regularization of the thresholds at each node, thus the spread of the data doesn't matter and the algorithm isn't effected by scale. 

### 4.4 Hyperparameter optimization
rubric={points:10}

In this exercise, you'll carry out hyperparameter optimization for the hyperparameter `C` of SVC RBF classifier. In practice you'll carry out hyperparameter optimization for all different hyperparameters for the most promising classifiers. For the purpose of this assignment, we'll only do it for the `SVC` classifier with one hyperparameter: `C`. 

**Your tasks:**

1. For each `C` value in the `param_grid` in the starter code below: 
    - Create a pipeline object with two steps: preprocessor from 3.1 and `SVC` classifier with the value of `C`.
    - Carry out 5-fold cross validation with the pipeline.  
    - Store the results in `results_dict` and display results as a pandas DataFrame. 
2. Which hyperparameter value seems to be performing the best? Is it different than the default value for the hyperparameter used by `scikit-learn`? 

> Note: Running this might take a while. 

In [ ]:
param_grid = {"C": np.logspace(-2, 2, 4)}
for c in param_grid["C"]: 
    hyperparam_pipe = make_pipeline(preprocessor, SVC(C=c))
    results_dict["RBF SVM" + ": " + str(c)] = mean_std_cross_val_scores(
    hyperparam_pipe, X_train, y_train, return_train_score=True)

pd.DataFrame(results_dict).T

### The best hyperparameter seems to be either C≈4.64 or C=100. C is a regularization error that controls the tradeoff between low training error and low testing error, and in either case we are allowing for fewer outliers than the default value for the hyperparameter used by scikit-learn. Since the test scores are the same, I will use C≈4.64 because Iam able to train the model faster. 

<br><br><br><br>

## Exercise 5: Evaluating on the test set <a name="5"></a>
<hr>

Now that we have a best performing model, it's time to assess our model on the set aside test set. In this exercise you'll examine whether the results you obtained using cross-validation on the train set are consistent with the results on the test set. 

### 5.1 Scoring on the unseen test set 
rubric={points:10}

**Your tasks:**

1. Train the best performing model on the entire training set. 
2. Report the results of this model on `X_test`. 
3. Are the cross-validation results and test results consistent? 

In [ ]:
model = make_pipeline(preprocessor, SVC(C=param_grid["C"][2]))

model.fit(X_train, y_train)

model.score(X_test, y_test)


### The test score is ≈ 0.8513, which is very close to the cross-validation results. This indicates that the model we have trained had a good trade-off between bias (underfitting) and variance (overfitting). 

<br><br><br><br>

## Submission instructions 

**PLEASE READ:** When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission. 

Congratulations on finishing the homework! This was a tricky one but I hope you are feeling good after working on it. You are now ready to build a simple supervised machine learning pipeline on real-world datasets! Well done :clap:! 

![](eva-well-done.png)

